vendor/brave-ios/Ads/BATBraveAds.mm

```diff
@@ -512,6 +516,219 @@ - (void)URLRequest:(const std::string &)url headers:(const std::vector<std::stri
+    formatter.dateStyle = NSDateFormatterFullStyle;
+    formatter.timeStyle = NSDateFormatterFullStyle;
+    BLOG(1, @"Update user models on %@", [formatter stringFromDate:[[NSDate date] dateByAddingTimeInterval:time_offset]]);
+  }];
+
+  [self.updateUserModelTimer fire];
+}
+
+- (void)updateUserModels
+{
+  NSDictionary *dict = [self userModelMetadata];
+  BLOG(1, @"Updating user models");
+
+  for (NSString *key in dict) {
+    BLOG(1, @"Checking %@ user model for updates", key);
+
+    const auto __weak weakSelf = self;
+    [self downloadUserModelForId:std::string([key UTF8String]) completion:^(bool success, bool should_retry) {
+      const auto strongSelf = weakSelf;
+      if (!strongSelf) { return; }
+
+      if (!success) {
+        return;
+      }
+
+      strongSelf->ads->OnUserModelUpdated(std::string([key UTF8String]));
+    }];
+  }
+}
+
+- (void)downloadUserModelForId:(const std::string &)id completion:(void (^)(bool, bool))completion
+{
+  BLOG(1, @"Downloading %@ user model manifest", [NSString stringWithUTF8String:id.c_str()]);
+
+  const auto __weak weakSelf = self;
+
+  NSString *baseUrl;
+  if (ads::_environment == ads::Environment::PRODUCTION) {
+    baseUrl = @"https://brave-user-model-installer-input.s3.brave.com";
+  } else {
+    baseUrl = @"https://brave-user-model-installer-input-dev.s3.bravesoftware.com";
+  }
+
+  NSString *bridgedId = [NSString stringWithUTF8String:id.c_str()];
+
+  NSString *userModelPath = self.userModelPaths[bridgedId] ?: @"";
+  baseUrl = [baseUrl stringByAppendingPathComponent:userModelPath];
+
+  NSString *manifestUrl = [baseUrl stringByAppendingPathComponent:@"models.json"];
+  return [self.commonOps loadURLRequest:std::string([manifestUrl UTF8String]) headers:{} content:"" content_type:"" method:"GET" callback:^(int statusCode, const std::string &response, const std::map<std::string, std::string> &headers) {
+    const auto strongSelf = weakSelf;
+    if (!strongSelf) { return; }
+
+    if (statusCode == 404) {
+      BLOG(1, @"%@ user model manifest not found", [NSString stringWithUTF8String:id.c_str()]);
+      completion(NO, NO);
+      return;
+    }
+
+    if (statusCode != 200) {
+      BLOG(1, @"Failed to download %@ user model manifest", [NSString stringWithUTF8String:id.c_str()]);
+      completion(NO, YES);
+      return;
+    }
+
+    NSData *data = [[NSString stringWithUTF8String:response.c_str()] dataUsingEncoding:NSUTF8StringEncoding];
+    NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
+    if (!dict) {
+      completion(NO, YES);
+      return;
+    }
+
+    NSNumber *schemaVersion = dict[@"schemaVersion"];
+    if ([schemaVersion intValue] != kCurrentUserModelManifestSchemaVersion) {
+      completion(NO, YES);
+      return;
+    }
+
+    NSArray *models = dict[@"models"];
+
+    for (NSDictionary *model in models) {
+      if (![model objectForKey:@"id"]) {
+        continue;
+      }
+      NSString *modelId = model[@"id"];
```

Comment:
Change this (and the ones below) to follow:

```
NSString *modelId = model[@"id"];
if (!modelId) {
  continue;
}
```

So you dont have to dupe the key and read from dict redundantly

Reply:
Fixed